packages/@react-aria/grid/src/useGridCell.ts

```diff
@@ -205,18 +249,101 @@ export function useGridCell<T, C extends GridCollection<T>>(props: GridCellProps
         // to the cell above/below instead. We need to re-dispatch the event from a higher parent so it still
         // bubbles and gets handled by useSelectableCollection.
-        if (!e.altKey && ref.current.contains(e.target as Element)) {
+        if (keyboardNavigationBehavior === 'arrow' && !e.altKey && ref.current.contains(e.target as Element)) {
           e.stopPropagation();
           e.preventDefault();
-          ref.current.parentElement.dispatchEvent(
-            new KeyboardEvent(e.nativeEvent.type, e.nativeEvent)
-          );
+
+          if (isEditing) {
+            ref.current.dispatchEvent(
+              new KeyboardEvent(e.nativeEvent.type, {...e.nativeEvent, key: e.key === 'ArrowUp' ? 'ArrowLeft' : 'ArrowRight'})
+            );
+          } else {
+            ref.current.parentElement.dispatchEvent(
+              new KeyboardEvent(e.nativeEvent.type, e.nativeEvent)
+            );
+          }
+        }
+        break;
+      }
+      case 'F2':
+      case 'Enter': {
+        if (isEditable) {
+          let focusable = walker.nextNode() as FocusableElement;
+
+          if (focusable && document.activeElement === ref.current) {
+            state.setEditingKey(node.key);
+            focusSafely(focusable);
+          }
+        }
+
+        if (isEditable && state.editingKey && (isSingleLineTextInput(document.activeElement) || e.key === 'F2')) {
+          state.setEditingKey(null);
+          focusSafely(ref.current);
+        }
+        break;
+      }
+      case 'Escape': {
+        if (state.editingKey || keyboardNavigationBehavior === 'tab') {
+          state.setEditingKey(null);
+          focusSafely(ref.current);
+        }
+        break;
+      }
+      case 'Tab': {
+        if (keyboardNavigationBehavior === 'tab') {
+          // If there is another focusable element within this item, stop propagation so the tab key
+          // is handled by the browser and not by useSelectableCollection (which would take us out of the list).
+          let walker = getFocusableTreeWalker(ref.current, {tabbable: true});
+          walker.currentNode = document.activeElement;
+          let next = e.shiftKey ? walker.previousNode() : walker.nextNode();
+
+          if (next || isEditing) {
+            e.stopPropagation();
+
+            if (isEditable && document.activeElement === ref.current) {
+              state.setEditingKey(node.key); // Do we want this?
+            }
+
+            let [start, end] = wrap(walker);
+
+            if (isEditing && !e.shiftKey && document.activeElement === end) {
+              e.preventDefault();
+              focusSafely(start);
+            }
+
+            if (isEditing && e.shiftKey && document.activeElement === start) {
+              e.preventDefault();
+              focusSafely(end);
+            }
+          }
+        }
+        break;
+      }
+      default: {
+        if (document.activeElement === ref.current && /^[a-zA-Z0-9]$/.test(e.key)) {
```

Comment:
Known issue: Entering an editable cell should be preferred over grid navigation, when node textValue is matched to a row.

Reply:
I believe the idea was so that the user could use typeahead to quickly navigate to a new row regardless of the cell they were in, but its been a while so I'll have to dig for any discussions on this. Personally, I feel like enforcing that entering a cell via Tab/Enter is required for editing the cell's content is alright, but not too strong of a stance tbh